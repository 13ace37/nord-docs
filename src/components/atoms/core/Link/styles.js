/*
 * Copyright (C) 2018-present Arctic Ice Studio <development@arcticicestudio.com>
 * Copyright (C) 2018-present Sven Greb <development@svengreb.de>
 *
 * Project:    Nord Docs
 * Repository: https://github.com/arcticicestudio/nord-docs
 * License:    MIT
 */

/**
 * @file Provides multiple styles for the `Link` component.
 * @author Arctic Ice Studio <development@arcticicestudio.com>
 * @author Sven Greb <development@svengreb.de>
 * @since 0.5.0
 */

import { css } from "styled-components";
import { lighten, rgba } from "polished";

import { colors, themedMode, MODE_BRIGHT_SNOW_FLURRY, MODE_DARK_NIGHT_FROST } from "styles/theme";

const backgroundColorHover = themedMode({
  [MODE_BRIGHT_SNOW_FLURRY]: rgba(colors.nord6, 0.45),
  [MODE_DARK_NIGHT_FROST]: rgba(colors.nord3, 0.8)
});

const fontColor = themedMode({
  [MODE_BRIGHT_SNOW_FLURRY]: colors.nord10,
  [MODE_DARK_NIGHT_FROST]: colors.nord8
});

const fontColorDecent = themedMode({
  [MODE_BRIGHT_SNOW_FLURRY]: colors.nord0,
  [MODE_DARK_NIGHT_FROST]: lighten(0.2, colors.font.base[MODE_DARK_NIGHT_FROST])
});

const fontColorHover = themedMode({
  [MODE_BRIGHT_SNOW_FLURRY]: colors.nord10
});

const fontColorHoverMinimal = themedMode({
  [MODE_BRIGHT_SNOW_FLURRY]: colors.nord0,
  [MODE_DARK_NIGHT_FROST]: lighten(0.2, colors.font.base[MODE_DARK_NIGHT_FROST])
});

const calmly = css`
  border-radius: 0.25em;
  color: ${fontColor};
  transition: color 0.2s ease-in-out, background-color 0.2s ease-in-out;

  &:hover,
  &:active,
  &:focus {
    background-color: ${backgroundColorHover};
    color: ${fontColorHover};
  }
`;

const decent = css`
  border-bottom: 1px solid ${fontColor};
  color: ${fontColorDecent};
  transition: color 0.2s ease-in-out;

  &:hover,
  &:active,
  &:focus {
    color: ${fontColor};
  }
`;

const minimal = css`
  border-bottom: 1px solid ${fontColor};
  transition: border-bottom-color 0.2s ease-in-out;

  &:hover,
  &:active,
  &:focus {
    border-bottom-color: ${fontColorHoverMinimal};
  }
`;

export { calmly, decent, minimal };
