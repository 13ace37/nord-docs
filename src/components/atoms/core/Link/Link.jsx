/*
 * Copyright (C) 2018-present Arctic Ice Studio <development@arcticicestudio.com>
 * Copyright (C) 2018-present Sven Greb <development@svengreb.de>
 *
 * Project:    Nord Docs
 * Repository: https://github.com/arcticicestudio/nord-docs
 * License:    MIT
 */

import PropTypes from "prop-types";
import styled from "styled-components";

import { A } from "atoms/core/html-elements";

import { calmly, decent, minimal } from "./styles";

const variants = {
  calmly,
  decent,
  minimal
};

/**
 * A wrapper for the base HTML component `A` with multiple styles that can be selected through the `variant` prop.
 *
 * @author Arctic Ice Studio <development@arcticicestudio.com>
 * @author Sven Greb <development@svengreb.de>
 * @since 0.5.0
 */
const Link = styled(A)`
  ${({ variant }) => variants[variant]};
`;

Link.propTypes = {
  variant: PropTypes.oneOf(Object.keys(variants))
};

Link.defaultProps = {
  variant: "calmly"
};

export default Link;
