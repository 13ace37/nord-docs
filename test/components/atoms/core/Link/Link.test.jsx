/*
 * Copyright (C) 2018-present Arctic Ice Studio <development@arcticicestudio.com>
 * Copyright (C) 2018-present Sven Greb <development@svengreb.de>
 *
 * Project:    Nord Docs
 * Repository: https://github.com/arcticicestudio/nord-docs
 * License:    MIT
 */

import React from "react";

import { renderWithTheme } from "nord-docs-test-utils";
import Link from "atoms/core/Link";

describe("theme styles", () => {
  test("matches the snapshot with default variant", () => {
    const { container } = renderWithTheme(<Link>Nord</Link>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("matches the snapshot with `calmly` variant", () => {
    const { container } = renderWithTheme(<Link variant="calmly">Nord</Link>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("matches the snapshot with `decent` variant", () => {
    const { container } = renderWithTheme(<Link variant="decent">Nord</Link>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("matches the snapshot with `minimal` variant", () => {
    const { container } = renderWithTheme(<Link variant="minimal">Nord</Link>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
