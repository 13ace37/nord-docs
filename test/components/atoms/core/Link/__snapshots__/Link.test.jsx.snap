// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`theme styles matches the snapshot with \`calmly\` variant 1`] = `
.c1 {
  color: inherit;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:active,
.c1:focus,
.c1:hover,
.c1:visited {
  outline: none;
}

.c0 {
  border-radius: 0.25em;
  color: #5e81ac;
  -webkit-transition: color 0.2s ease-in-out,background-color 0.2s ease-in-out;
  transition: color 0.2s ease-in-out,background-color 0.2s ease-in-out;
}

.c0:hover,
.c0:active,
.c0:focus {
  background-color: rgba(236,239,244,0.45);
  color: #5e81ac;
}

<a
  class="c0 c1"
  href=""
>
  Nord
</a>
`;

exports[`theme styles matches the snapshot with \`decent\` variant 1`] = `
.c1 {
  color: inherit;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:active,
.c1:focus,
.c1:hover,
.c1:visited {
  outline: none;
}

.c0 {
  border-bottom: 1px solid #5e81ac;
  color: #2e3440;
  -webkit-transition: color 0.2s ease-in-out;
  transition: color 0.2s ease-in-out;
}

.c0:hover,
.c0:active,
.c0:focus {
  color: #5e81ac;
}

<a
  class="c0 c1"
  href=""
>
  Nord
</a>
`;

exports[`theme styles matches the snapshot with \`minimal\` variant 1`] = `
.c1 {
  color: inherit;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:active,
.c1:focus,
.c1:hover,
.c1:visited {
  outline: none;
}

.c0 {
  border-bottom: 1px solid #5e81ac;
  -webkit-transition: border-bottom-color 0.2s ease-in-out;
  transition: border-bottom-color 0.2s ease-in-out;
}

.c0:hover,
.c0:active,
.c0:focus {
  border-bottom-color: #2e3440;
}

<a
  class="c0 c1"
  href=""
>
  Nord
</a>
`;

exports[`theme styles matches the snapshot with default variant 1`] = `
.c1 {
  color: inherit;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:active,
.c1:focus,
.c1:hover,
.c1:visited {
  outline: none;
}

.c0 {
  border-radius: 0.25em;
  color: #5e81ac;
  -webkit-transition: color 0.2s ease-in-out,background-color 0.2s ease-in-out;
  transition: color 0.2s ease-in-out,background-color 0.2s ease-in-out;
}

.c0:hover,
.c0:active,
.c0:focus {
  background-color: rgba(236,239,244,0.45);
  color: #5e81ac;
}

<a
  class="c0 c1"
  href=""
>
  Nord
</a>
`;
